From 81fef0288580b877a5f0a407812669653d02221d Mon Sep 17 00:00:00 2001
From: Guillem Jover <guillem@hadrons.org>
Date: Sun, 6 Dec 2009 05:48:41 +0100
Subject: [PATCH] Split configuration file loading into a new function

* syslogd/syslogd.c (load_conffile): New declaration.
(init): Move configuration file loading into ...
(load_conffile): ... here. New function.
---
 src/syslogd.c | 120 ++++++++++++++++++++++++++++++++--------------------------
 1 file changed, 66 insertions(+), 54 deletions(-)

diff --git a/src/syslogd.c b/src/syslogd.c
index fc633a8..0a4d698 100644
--- a/src/syslogd.c
+++ b/src/syslogd.c
@@ -258,6 +258,7 @@ void doexit (int);
 void domark (int);
 void find_inet_port (const char *);
 void fprintlog (struct filed *, const char *, int, const char *);
+static int load_conffile (const char *, struct filed **);
 void init (int);
 void logerror (const char *);
 void logmsg (int, const char *, const char *, int);
@@ -1796,12 +1797,11 @@ die (int signo)
   exit (EXIT_SUCCESS);
 }
 
-/* INIT -- Initialize syslogd from configuration table.  */
-void
-init (int signo _GL_UNUSED_PARAMETER)
+static int
+load_conffile (const char *filename, struct filed **nextp)
 {
   FILE *cf;
-  struct filed *f, *next, **nextp;
+  struct filed *f;
   char *p;
 #ifndef LINE_MAX
 # define LINE_MAX 2048
@@ -1811,64 +1811,19 @@ init (int signo _GL_UNUSED_PARAMETER)
   char *cline;
   int cont_line = 0;
 
-  dbg_printf ("init\n");
-
-  /* Close all open log files.  */
-  Initialized = 0;
-  for (f = Files; f != NULL; f = next)
-    {
-      int j;
-
-      /* Flush any pending output.  */
-      if (f->f_prevcount)
-	fprintlog (f, LocalHostName, 0, (char *) NULL);
-
-      switch (f->f_type)
-	{
-	case F_FILE:
-	case F_TTY:
-	case F_CONSOLE:
-	case F_PIPE:
-	  free (f->f_un.f_fname);
-	  close (f->f_file);
-	  break;
-	case F_FORW:
-	case F_FORW_SUSP:
-	case F_FORW_UNKN:
-	  free (f->f_un.f_forw.f_hname);
-	  break;
-	case F_USERS:
-	  for (j = 0; j < f->f_un.f_user.f_nusers; ++j)
-	    free (f->f_un.f_user.f_unames[j]);
-	  free (f->f_un.f_user.f_unames);
-	  break;
-	}
-      free (f->f_progname);
-      free (f->f_prevhost);
-      next = f->f_next;
-      free (f);
-    }
-  Files = NULL;
-  nextp = &Files;
-
-  facilities_seen = 0;
-
   /* Open the configuration file.  */
-  cf = fopen (ConfFile, "r");
+  cf = fopen (filename, "r");
   if (cf == NULL)
     {
-      dbg_printf ("cannot open %s\n", ConfFile);
+      dbg_printf ("cannot open %s\n", filename);
       *nextp = (struct filed *) calloc (1, sizeof (*f));
       cfline ("*.ERR\t" PATH_CONSOLE, *nextp);
       (*nextp)->f_next = (struct filed *) calloc (1, sizeof (*f));
       cfline ("*.PANIC\t*", (*nextp)->f_next);
       Initialized = 1;
-      return;
+      return 1;
     }
 
-  /* Foreach line in the conf table, open that file.  */
-  f = NULL;
-
   /* Allocate a buffer for line parsing.  */
   cbuf = malloc (line_max);
   if (cbuf == NULL)
@@ -1876,7 +1831,7 @@ init (int signo _GL_UNUSED_PARAMETER)
       /* There is no graceful recovery here.  */
       dbg_printf ("cannot allocate space for configuration\n");
       fclose (cf);
-      return;
+      return 0;
     }
   cline = cbuf;
 
@@ -1921,7 +1876,7 @@ init (int signo _GL_UNUSED_PARAMETER)
 	      dbg_printf ("cannot allocate space configuration\n");
 	      fclose (cf);
 	      free (cbuf);
-	      return;
+	      return 0;
 	    }
 	  else
 	    cbuf = tmp;
@@ -2014,6 +1969,63 @@ init (int signo _GL_UNUSED_PARAMETER)
   fclose (cf);
   free (cbuf);
 
+  return 1;
+}
+
+/* INIT -- Initialize syslogd from configuration table.  */
+void
+init (int signo _GL_UNUSED_PARAMETER)
+{
+  struct filed *f, *next, **nextp;
+  struct servent *sp;
+
+  dbg_printf ("init\n");
+
+  /* Close all open log files.  */
+  Initialized = 0;
+  for (f = Files; f != NULL; f = next)
+    {
+      int j;
+
+      /* Flush any pending output.  */
+      if (f->f_prevcount)
+	fprintlog (f, LocalHostName, 0, (char *) NULL);
+
+      switch (f->f_type)
+	{
+	case F_FILE:
+	case F_TTY:
+	case F_CONSOLE:
+	case F_PIPE:
+	  free (f->f_un.f_fname);
+	  close (f->f_file);
+	  break;
+	case F_FORW:
+	case F_FORW_SUSP:
+	case F_FORW_UNKN:
+	  free (f->f_un.f_forw.f_hname);
+	  break;
+	case F_USERS:
+	  for (j = 0; j < f->f_un.f_user.f_nusers; ++j)
+	    free (f->f_un.f_user.f_unames[j]);
+	  free (f->f_un.f_user.f_unames);
+	  break;
+	}
+      free (f->f_progname);
+      free (f->f_prevhost);
+      next = f->f_next;
+      free (f);
+    }
+  Files = NULL;
+  nextp = &Files;
+
+  facilities_seen = 0;
+
+  /* Foreach line in the conf table, open that file.  */
+  f = NULL;
+
+  load_conffile (ConfFile, nextp);
+
   Initialized = 1;
 
   if (Debug)
-- 
1.8.3.rc1

